### PR TITLE
fix: workaround for mChrome empty space when navigating after address bar collapses

### PR DESCRIPTION
**Pull request by BeeMargarida**
**Motivation**

Related to https://github.com/react-navigation/react-navigation/issues/11274 and https://github.com/Expensify/App/issues/15749.

In mobile Chrome, the address bar collapses, which causes some problems: in an SPA and using a `flex: 1; height: 100%` style for the document, if the navigation is made with the address bar collapsed, it will show an empty space at the bottom of the screen (only visible if there are differences in background color). 
A possible fix is updating the body height using viewport height units based on the `window.innerHeight`, which should have the correct height taking into consideration the collapsed address bar. 

Resources:
- https://developer.chrome.com/blog/url-bar-resizing/
- https://css-tricks.com/the-trick-to-viewport-units-on-mobile/ 

**Test plan**

Run the example app in mobile chrome 

<details>
<summary>AuthFlow example code modified to test this exact scenario</summary>

```tsx
import { ParamListBase, useTheme } from '@react-navigation/native';
import {
  createStackNavigator,
  StackScreenProps,
} from '@react-navigation/stack';
import * as React from 'react';
import {
  ActivityIndicator,
  ScrollView,
  StyleSheet,
  TextInput,
  View,
} from 'react-native';
import { Button, Title } from 'react-native-paper';
import { SafeAreaProvider } from 'react-native-safe-area-context';

type AuthStackParams = {
  Home: undefined;
  SignIn: undefined;
  Chat: undefined;
};

const AUTH_CONTEXT_ERROR =
  'Authentication context not found. Have your wrapped your components with AuthContext.Consumer?';

const AuthContext = React.createContext<{
  isSignedIn: boolean;
  signIn: () => void;
  signOut: () => void;
}>({
  isSignedIn: false,
  signIn: () => {
    throw new Error(AUTH_CONTEXT_ERROR);
  },
  signOut: () => {
    throw new Error(AUTH_CONTEXT_ERROR);
  },
});

const SplashScreen = () => {
  const { colors } = useTheme();

  return (
    <View style={styles.content}>
      <ActivityIndicator color={colors.primary} />
    </View>
  );
};

const SignInScreen = ({
  navigation,
}: StackScreenProps<AuthStackParams, 'SignIn'>) => {
  const { signIn } = React.useContext(AuthContext);
  const { colors } = useTheme();

  return (
    <ScrollView>
      <View style={styles.content}>
        <TextInput
          placeholder="Username"
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <TextInput
          placeholder="Password"
          secureTextEntry
          style={[
            styles.input,
            { backgroundColor: colors.card, color: colors.text },
          ]}
        />
        <Button
          mode="contained"
          onPress={() => setTimeout(() => signIn(), 1000)}
          style={styles.button}
        >
          Sign in
        </Button>
        <Button
          onPress={() => navigation.navigate('Chat')}
          style={styles.button}
        >
          Go to Chat
        </Button>
      </View>
    </ScrollView>
  );
};

const HomeScreen = ({
  navigation,
}: StackScreenProps<AuthStackParams, 'Home'>) => {
  const { signOut } = React.useContext(AuthContext);

  return (
    <View style={styles.content}>
      <Title style={styles.text}>Signed in successfully 🎉</Title>
      <Button onPress={signOut} style={styles.button}>
        Sign out
      </Button>
      <Button onPress={() => navigation.navigate('Chat')} style={styles.button}>
        Go to Chat
      </Button>
    </View>
  );
};

const ChatScreen = ({ navigation }) => {
  const { isSignedIn, signIn, signOut } = React.useContext(AuthContext);

  React.useEffect(() => {
    const unsubscribe = navigation.addListener('focus', () => {
      if (typeof document === 'undefined' || !document.body) {
        // Only run when DOM is available
        return;
      }

      console.log('hererere', window.scrollY);

      document.getElementById('root')?.scrollIntoView();

      // window.scrollTo(0, 0);
    });

    // Return the function to unsubscribe from the event so it gets removed on unmount
    return unsubscribe;
  }, [navigation]);

  return (
    <View style={styles.content}>
      <Title style={styles.text}>What&apos;s up?</Title>
      <Title style={styles.text}>What&apos;s up?</Title>
      <Title style={styles.text}>What&apos;s up?</Title>
      <Title style={styles.text}>What&apos;s up?</Title>
      <Title style={{ position: 'absolute', bottom: 0 }}>Bottom</Title>
      {isSignedIn ? (
        <Button onPress={signOut} style={styles.button}>
          Sign out
        </Button>
      ) : (
        <Button onPress={signIn} style={styles.button}>
          Sign in
        </Button>
      )}
    </View>
  );
};

const SimpleStack = createStackNavigator<AuthStackParams>();

type State = {
  isLoading: boolean;
  isSignout: boolean;
  userToken: undefined | string;
};

type Action =
  | { type: 'RESTORE_TOKEN'; token: undefined | string }
  | { type: 'SIGN_IN'; token: string }
  | { type: 'SIGN_OUT' };

export function AuthFlow({ navigation }: StackScreenProps<ParamListBase>) {
  const [state, dispatch] = React.useReducer<React.Reducer<State, Action>>(
    (prevState, action) => {
      switch (action.type) {
        case 'RESTORE_TOKEN':
          return {
            ...prevState,
            userToken: action.token,
            isLoading: false,
          };
        case 'SIGN_IN':
          return {
            ...prevState,
            isSignout: false,
            userToken: action.token,
          };
        case 'SIGN_OUT':
          return {
            ...prevState,
            isSignout: true,
            userToken: undefined,
          };
      }
    },
    {
      isLoading: true,
      isSignout: false,
      userToken: undefined,
    }
  );

  React.useEffect(() => {
    const timer = setTimeout(() => {
      dispatch({ type: 'RESTORE_TOKEN', token: undefined });
    }, 1000);

    return () => clearTimeout(timer);
  }, []);

  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    });
  }, [navigation]);

  const isSignedIn = state.userToken !== undefined;

  const authContext = React.useMemo(
    () => ({
      isSignedIn,
      signIn: () => dispatch({ type: 'SIGN_IN', token: 'dummy-auth-token' }),
      signOut: () => dispatch({ type: 'SIGN_OUT' }),
    }),
    [isSignedIn]
  );

  if (state.isLoading) {
    return <SplashScreen />;
  }

  return (
    <SafeAreaProvider>
      <AuthContext.Provider value={authContext}>
        <SimpleStack.Navigator>
          {!isSignedIn ? (
            <SimpleStack.Screen
              name="SignIn"
              options={{
                title: 'Sign in',
                animationTypeForReplace: state.isSignout ? 'pop' : 'push',
                // cardStyle: { flex: 1 },
              }}
              component={SignInScreen}
            />
          ) : (
            <SimpleStack.Screen name="Home" component={HomeScreen} />
          )}
          <SimpleStack.Screen
            navigationKey={String(isSignedIn)}
            name="Chat"
            component={ChatScreen}
          />
        </SimpleStack.Navigator>
      </AuthContext.Provider>
    </SafeAreaProvider>
  );
}

const styles = StyleSheet.create({
  content: {
    flex: 1,
    padding: 16,
    justifyContent: 'center',
  },
  input: {
    margin: 8,
    padding: 10,
    borderRadius: 3,
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: 'rgba(0, 0, 0, 0.08)',
  },
  button: {
    margin: 8,
  },
  text: {
    textAlign: 'center',
    margin: 8,
  },
});

```
</details>

| Before | With the fix  |
| --------------- | --------------- |
| <video src="https://github.com/react-navigation/react-navigation/assets/25725586/1dc6e590-3d9b-44f3-9529-e4e232d0cbf7"> |  <video src="https://github.com/react-navigation/react-navigation/assets/25725586/db9f2071-e6a6-42f5-b26c-3e9e29833ffd"> |

This next video contains an example of the modified AuthFlow with the fix applied. As it can be seen, if there is bottom content, it will cut when the address bar comes back.

https://github.com/react-navigation/react-navigation/assets/25725586/72ba5b4a-95e0-4bfb-bb1c-ea70111bc334

### CODE REVIEW

**Comment by BeeMargarida** [comment]
Pinging about this, so that it does not get forgotten

File: packages/stack/src/views/Stack/CardSheet.tsx [hunk 1]
```diff
@@ -36,6 +36,18 @@ export const CardSheet = React.forwardRef<CardSheetRef, Props>(
       const width = document.body.clientWidth;
       const height = document.body.clientHeight;
 
+      // Workaround for mobile Chrome, necessary when a navigation happens
+      // when the address bar has already collapsed, which resulted in an
+      // empty space at the bottom of the page (matching the height of the
+      // address bar). To fix this, it's necessary to update the height of
+      // the DOM with the current height of the window.
+      const vh = window.innerHeight * 0.01;
```
Comment (osdnk) [inline packages/stack/src/views/Stack/CardSheet.tsx:44]: I wonder if we need to check if this happens on the web?

Comment (osdnk) [reply packages/stack/src/views/Stack/CardSheet.tsx:44]: ah right, we check it above, nvm

Comment (osdnk) [inline packages/stack/src/views/Stack/CardSheet.tsx:44]: can you maybe include the link to the CSS hack website?

**Review by osdnk — APPROVED** [review]
I confirm it works on my S22 Ultra and treat the issue was visible earlier. 

The issue doesn't seem tobe visible on any other browser (including chrome on iOS)

**Comment by osdnk** [comment]
If you want, I feel confident to go forward with the PR

File: packages/stack/src/views/Stack/CardSheet.tsx (hunk 1 above)
Comment (BeeMargarida) [reply packages/stack/src/views/Stack/CardSheet.tsx:45]: Yap, added 👍

File: packages/stack/src/views/Stack/CardSheet.tsx [hunk 2]
```diff
@@ -36,6 +36,19 @@ export const CardSheet = React.forwardRef<CardSheetRef, Props>(
...
+      // the DOM with the current height of the window.
+      // See https://css-tricks.com/the-trick-to-viewport-units-on-mobile/
+      const vh = window.innerHeight * 0.01;
+      document.documentElement.style.setProperty('--vh', `${vh}px`);
+      document.body.setAttribute(
+        'style',
+        `height: calc(var(--vh, 1vh) * 100);`
+      );
+
       setFill(width === layout.width && height === layout.height);
```
Comment (osdnk) [inline packages/stack/src/views/Stack/CardSheet.tsx:52]: This thing which I'm still worried about is that in some apps, authors set the `body`'s style to `height: 100%`, in which case the collapsing won't happen. Alternatively, you can render the stack in a smaller part of the page. Then our fix is no longer valid and may break some things (better not to hack, if not needed). 


```suggestion
      const fillHeight = height === layout.height;
      if (fillHeight) {
        const vh = window.innerHeight * 0.01;
        document.documentElement.style.setProperty('--vh', `${vh}px`);
        document.body.setAttribute(
          'style',
          `height: calc(var(--vh, 1vh) * 100);`
        );
        workaroundApplied.current = true;
      } else { 
        if (workaroundApplied.current) {
          // revert the fix if it was set
          // e.g., document.documentElement.style.removeProperty('--vh');
        }
      }

      setFill(width === layout.width && height === layout.height);
```

Comment (BeeMargarida) [reply packages/stack/src/views/Stack/CardSheet.tsx]: That makes sense, thank you 👍 I've applied the fix

**Pull request merged**